### PR TITLE
Cache for QuadGKJL

### PR DESCRIPTION
**Pull request by lxvm**
Hi, this pr makes the `IntegralCache` mechanism nontrivial and adds a working cache for the `QuadGKJL` algorithm. What makes this cache somewhat tricky is that it needs to be rebuilt eagerly, i.e. whenever the integrand or parameters are reset, we allocate a new cache since these changes could change the return type of the integrand. Consider the example below
```
using Integrals
f(x, p) = sin(x * p) # real-valued integrand
prob = IntegralProblem(f, 0, 1, 14.0)
cache = init(prob, QuadGKJL()) # cache for real values
solve!(cache) # u: 0.061661627270869046
g(x, p) = cis(x * p) # complex-valued integrand
cache = Integrals.set_f(cache, g) # new cache for complex values
solve!(cache) # u: 0.07075766826391927 + 0.061661627270869046im
```
If we could let the caller specify that the cache need not be changed, since they guarantee that the return type of the integral is unchanged, I think this would let us reuse the allocated cache across multiple calls. Otherwise I think we cannot safely reuse an immutable, concretely-typed cache without some guarantee of type stability when changing the integrand or input parameters.

In summary, the following cache-resetting methods were added
- `set_f(cache, f, [nout])`: rebuilds cache
- `set_lb(cache, lb)`: does not rebuild cache
- `set_ub(cache, ub)`: does not rebuild cache
- `set_p(cache, p)`: rebuilds cache

I also made an effort to get the AD to work with these changes. ForwardDiff.jl should still be working, however I still run into  https://github.com/SciML/Integrals.jl/issues/99 with Zygote.jl.

I would appreciate any feedback on the design and will gladly implement those changes!

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
In general it looks good. Any new interface like this needs to make sure it's documented. 

We might want to consider making the cache mutable. This was recently changed in LinearSolve. It was done there mostly for correctness reasons https://github.com/SciML/LinearSolve.jl/pull/303, but it means that Integrals is the only thing on an immutable cache now. There may still be reasons for it, but it is the odd one out now.

**Comment by ChrisRackauckas** [comment]
@ArnoStrouwen is there a reason this never merged?

**Comment by ChrisRackauckas** [comment]
I think an immutable cache style with an `init` function and `solve!`, matching the rest of SciML, would be the most desirable since it would then be uniform to all of the other solvers and not an odd ball, and we have seen a few cases where that improves performance and correctness more. 

>  changing the integrand in the cache would change the cache type?

That's expected. I think for integrals the main use case for the cache would be changing parameters? If we setup function wrappers this will not be the case anyways, since then all functions are the same type.

**Comment by ChrisRackauckas** [comment]
> but if you meant using https://github.com/yuyichao/FunctionWrappers.jl then my question about using Base.promote_op to infer the integrand's type could be answered.

I meant this, and instead of `promote_op` it would need `Core.return_type`, or we change the interface to have that passed in (preferred) like is being discussed in https://github.com/SciML/Integrals.jl/issues/169

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
This is fantastic. Thanks so much for taking it to completion!
